keras_nlp/src/layers/modeling/anchor_generator.py

```diff
@@ -0,0 +1,305 @@
+# Copyright 2024 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+
+import keras
+from keras import ops
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.bounding_box import converters
+
+
+@keras_nlp_export("keras_nlp.layers.AnchorGenerator")
+class AnchorGenerator(keras.layers.Layer):
```

Comment:
`AnchorGenerator`, `NonMaxSupression` and `BoxMatcher `layers are used in all Object Detection Models.

* YOLO
* Faster R-CNN
* RetinaNet
* Mas R-CNN

If any extra Object Detection Models are added in future these are mandatory layers across all these models. So added as part of layers.

If these layers are available from core directly we can isolate these to only particular model dir `RetinaNet` for now.

Reply:
SG! Well I think adding these in a separate PR will definitely be easier, and along with a model should help us suss out the task setup. I think we probably need so more substantial rewrites before we exposed these. In particular:

- I doubt we want to rely on raggeds as much. Let's write this the "jax and torch" way. So if we have ragged inputs, we probably want to convert them to padded dense inputs as a first step.
- We should use common test cases in the base test case, like our other layers.
- We should rewrite docstrings, etc to match local style.

The first of these is probably the trickiest, but it will be easiest to do when we have a actual model to work with.

Reply:
@mattdangerw,

So in the next PR make changes as per "jax and torch" for layers and write these layers as of now at model directory.

I will try to write one layer at a time in the PR and follow the points who have mentioned.

Thanks!